src/util/psbt/map/input.rs

```diff
@@ -0,0 +1,202 @@
+    /// other scripts necessary for this input to pass validation.
+    pub final_script_witness: Option<Vec<Vec<u8>>>,
+    /// Unknown key-value pairs for this input.
+    pub unknown: HashMap<raw::Key, Vec<u8>>,
+}
+
+impl Map for Input {
+    fn insert_pair(&mut self, pair: raw::Pair) -> Result<(), encode::Error> {
+        let raw::Pair {
+            key: raw_key,
+            value: raw_value,
+        } = pair;
+
+        match raw_key.type_value {
+            0u8 => {
+                impl_psbt_insert_pair! {
+                    self.non_witness_utxo <= <raw_key: _>|<raw_value: Transaction>
+                }
+            }
+            1u8 => {
+                impl_psbt_insert_pair! {
+                    self.witness_utxo <= <raw_key: _>|<raw_value: TxOut>
+                }
+            }
+            3u8 => {
+                impl_psbt_insert_pair! {
+                    self.sighash_type <= <raw_key: _>|<raw_value: SigHashType>
+                }
+            }
+            4u8 => {
+                impl_psbt_insert_pair! {
+                    self.redeem_script <= <raw_key: _>|<raw_value: Script>
+                }
+            }
+            5u8 => {
+                impl_psbt_insert_pair! {
+                    self.witness_script <= <raw_key: _>|<raw_value: Script>
+                }
+            }
+            7u8 => {
+                impl_psbt_insert_pair! {
+                    self.final_script_sig <= <raw_key: _>|<raw_value: Script>
+                }
+            }
+            8u8 => {
+                impl_psbt_insert_pair! {
+                    self.final_script_witness <= <raw_key: _>|<raw_value: Vec<Vec<u8>>>
+                }
+            }
+            2u8 => {
+                impl_psbt_insert_pair! {
+                    self.partial_sigs <= <raw_key: PublicKey>|<raw_value: Vec<u8>>
+                }
+            }
+            6u8 => {
+                impl_psbt_insert_pair! {
+                    self.hd_keypaths <= <raw_key: PublicKey>|<raw_value: (Fingerprint, DerivationPath)>
+                }
+            }
+            _ => {
+                if self.unknown.contains_key(&raw_key) {
+                    return Err(Error::DuplicateKey(raw_key).into());
+                } else {
+                    self.unknown.insert(raw_key, raw_value);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_pairs(&self) -> Result<Vec<raw::Pair>, encode::Error> {
+        let mut rv: Vec<raw::Pair> = Default::default();
+
+        if let Some(_) = self.non_witness_utxo {
+            impl_psbt_get_pair! {
+                rv.push(self.non_witness_utxo as <0u8, _>|<Transaction>)
+            }
+        } else {
+            impl_psbt_get_pair! {
+                rv.push(self.witness_utxo as <1u8, _>|<TxOut>)
+            }
+        }
+
+        if self.final_script_sig.is_none() && self.final_script_witness.is_none() {
```

Comment:
From the BIP:
> The Input Finalizer must only accept a PSBT. For each input, the Input Finalizer determines if the input has enough data to pass validation. If it does, it must construct the scriptSig and scriptWitness and place them into the input key-value map. All other data except the UTXO and unknown fields in the input key-value map should be cleared from the PSBT. The UTXO should be kept to allow Transaction Extractors to verify the final network serialized transaction.

Perhaps the function could be renamed in the future, it's more providing an intermediary representation where everything will be serialized rather than just spitting out all the `Pair`s (which is what `get_pairs` seems to imply).

Reply:
Yeah I think `get_pairs` doesn't qualify as the "Input Finalizer" from the spec.

I'd be more in favor of an additional method `finalize` that strips all no-longer-relevant information from it. Now this would break round-tripping PSBT structs that have redundant information as `get_pairs` is used for serializing.

Reply:
I agree with @stevenroose. These `if` statements also prevent PSBTs from round-tripping which makes fuzzing harder.

Reply:
@stevenroose @apoelstra How about we remove these `if` statements for now, and in a future PR introduce `finalize`?